test/test_codegen.py

```diff
@@ -174,37 +188,65 @@ def test_get_temperature(mechname):
+                                       ("sanDiego", "H2")])
+def test_kinetics(mechname, fuel):
     """This function tests that pyrometheus-generated code
-    computes thermodynamic properties c_p, s_R, h_RT, and k_eq
-    correctly by comparing against Cantera"""
-    # Create Cantera and pyrometheus objects
+    computes the right rates of progress for given temperature
+    and composition"""
     sol = ct.Solution(f"{mechname}.cti", "gas")
+    code = pyro.gen_thermochem_code(sol)
+    mechfile = open(f"{mechname}.py", "w")
+    print(code, file=mechfile)
+    mechfile.close()
+
     ptk = pyro.get_thermochem_class(sol)()
-    # Loop over temperatures
-    temp = np.linspace(500.0, 3000.0, 10)
-    for t in temp:
 
-        # Get properties from pyrometheus
-        cp_pm = ptk.get_species_specific_heats_R(t)
-        s_pm = ptk.get_species_entropies_R(t)
-        h_pm = ptk.get_species_enthalpies_RT(t)
-        keq_pm = 1 / np.exp(ptk.get_equilibrium_constants(t))
-        # Set state in cantera for comparison
-        sol.TP = t, ct.one_atm
-        keq_ct = sol.equilibrium_constants
-        # Compare properties
-        cp_err = np.abs(cp_pm - sol.standard_cp_R).max()
-        s_err = np.abs(s_pm - sol.standard_entropies_R).max()
-        h_err = np.abs(h_pm - sol.standard_enthalpies_RT).max()
-        keq_err = np.abs((keq_pm - keq_ct) / keq_ct)
-        assert cp_err < 1.0e-13
-        assert s_err < 1.0e-13
-        assert h_err < 1.0e-13
-        for i, r in enumerate(sol.reactions()):
-            if r.reversible:
-                assert keq_err[i] < 1.0e-13
+    # Homogeneous reactor to get test data
+    init_temperature = 1500.0
+    equiv_ratio = 1.0
+    ox_di_ratio = 0.21
+    stoich_ratio = 3.0  # 0.5
+    i_fu = sol.species_index(f"{fuel}")
+    i_ox = sol.species_index("O2")
+    i_di = sol.species_index("N2")
+    x = np.zeros(ptk.num_species)
+    x[i_fu] = (ox_di_ratio*equiv_ratio)/(stoich_ratio+ox_di_ratio*equiv_ratio)
+    x[i_ox] = stoich_ratio*x[i_fu]/equiv_ratio
+    x[i_di] = (1.0-ox_di_ratio)*x[i_ox]/ox_di_ratio
+    sol.TPX = init_temperature, ct.one_atm, x
+    reactor = ct.IdealGasConstPressureReactor(sol)
+    sim = ct.ReactorNet([reactor])
+
+    time = 0.0
+    for step in range(50):
+        time += 1.0e-7  # 1.0e-6 for H2
+        sim.advance(time)
+        # Cantera kinetics
+        r_ct = reactor.kinetics.net_rates_of_progress
+        omega_ct = reactor.kinetics.net_production_rates
+        # Get state from Cantera
+        temp = reactor.T
+        rho = reactor.density
+        y = np.where(reactor.Y > 0, reactor.Y, 0)
+        # Prometheus kinetics
+        c = ptk.get_concentrations(rho, y)
+        r_pm = ptk.get_net_rates_of_progress(temp, c)
+        omega_pm = ptk.get_net_production_rates(rho, temp, y)
+        # Print
+        err_r = np.linalg.norm((r_ct-r_pm), np.inf)
+        err_omega = np.linalg.norm((omega_ct[0:-1]-omega_pm[0:-1]), np.inf)
```

Comment:
Comment on why the last element of the array is being left out.

Reply:
I don't know.  Alternative answer: these are not the array elements you are looking for. :jedi-hand-wave: